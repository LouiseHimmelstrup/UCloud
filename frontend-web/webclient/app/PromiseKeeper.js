export default class PromiseKeeper {
    constructor() {
        this.promises = [];
    }

    _cleanup() {
        this.promises = this.promises.filter(it => !it.isComplete);
    }

    makeCancelable(promise) {
        let hasCanceled_ = false;
        let cancelablePromise = {};
        cancelablePromise.promise = new Promise((resolve, reject) => {
                promise.then(
                    val => {
                        cancelablePromise.isComplete = true;
                        this._cleanup();
                        if (hasCanceled_) reject({isCanceled: true}); else resolve(val);
                    },
                    error => {
                        cancelablePromise.isComplete = true;
                        this._cleanup();
                        hasCanceled_ ? reject({isCanceled: true}) : reject(error)
                    }
                );
            }
        );
        cancelablePromise.cancel = () => hasCanceled_ = true;
        this.promises.push(cancelablePromise);
        return cancelablePromise;
    };

    /**
     *  Cancels all promises stored in the promise keeper.
     *  The held promises are cleared from the keeper as they are cancelled and no longer have any function
     */
    cancelPromises() {
        this.promises.forEach((it) => {
            it.cancel();
        });
        this.promises = [];
    }
}