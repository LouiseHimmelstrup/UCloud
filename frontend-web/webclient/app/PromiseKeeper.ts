export default class PromiseKeeper {
    promises: Array<CancelablePromise<any>>
    constructor() {
        this.promises = [];
    }

    _cleanup(): void {
        this.promises = this.promises.filter(it => !it.isComplete);
    }

    makeCancelable<T>(promise: Promise<T>): CancelablePromise<T> {
        let hasCanceled_ = false;
        let cancelablePromise = new CancelablePromise<T>();
        cancelablePromise.promise = new Promise((resolve: any, reject: any) => {
            promise.then(
                val => {
                    cancelablePromise.isComplete = true;
                    this._cleanup();
                    hasCanceled_ ? reject({ isCanceled: true }) : resolve(val);
                },
                error => {
                    cancelablePromise.isComplete = true;
                    this._cleanup();
                    hasCanceled_ ? reject({ isCanceled: true }) : reject(error)
                }
            );
        });
        cancelablePromise.cancel = () => cancelablePromise.hasCanceled_ = true;
        this.promises.push(cancelablePromise);
        return cancelablePromise;
    };

    /**
     *  Cancels all promises stored in the promise keeper.
     *  The held promises are cleared from the keeper as they are cancelled and no longer have any function
     */
    cancelPromises(): void {
        this.promises.forEach((it) => it.cancel());
        this.promises = [];
    }
}

class CancelablePromise<T> {
    isComplete: boolean
    promise: Promise<T>
    cancel: VoidFunction
    hasCanceled_: boolean
    constructor() { }
}