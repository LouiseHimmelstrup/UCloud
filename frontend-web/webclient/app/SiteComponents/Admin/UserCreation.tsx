import * as React from "react";
import { Form, Button, Input, Message } from "semantic-ui-react";


interface UserCreationState {
    submitted: boolean
    username: string
    password: string
    repeatedPassword: string
    passwordError: boolean
    usernameError: boolean
}

type UserCreationField = keyof UserCreationState;

class UserCreation extends React.Component<{}, UserCreationState> {
    constructor(props) {
        super(props);
        this.state = {
            submitted: false,
            username: "",
            password: "",
            repeatedPassword: "",
            passwordError: false,
            usernameError: false
        }
    }

    updateFields(field: UserCreationField, value: string) {
        const state = { ...this.state }
        state[field] = value;
        state.usernameError = false;
        state.passwordError = false;
        this.setState(() => state);
    }

    submit() {
        let error = false;
        const {username, password, repeatedPassword } = this.state;
        if (!username) {
            this.setState(() => ({ usernameError: true }));
            error = true;
        }
        if (!password || password !== repeatedPassword) {
            this.setState(() => ({ passwordError: true }));
            error = true;
        }
        if (!error) {
            // submit
        }


    }

    render() {
        return (
            <React.StrictMode>
                <Form>
                    <Form.Field error={this.state.usernameError}>
                        <label>Username</label>
                        <Input value={this.state.username} onChange={(_e, { value }) => this.updateFields("username", value)} placeholder="Username..." />
                    </Form.Field>
                    <Form.Field error={this.state.passwordError}>
                        <label>Password</label>
                        <Input value={this.state.password} type="password" onChange={(_e, { value }) => this.updateFields("password", value)} placeholder="Password..." />
                    </Form.Field>
                    <Form.Field error={this.state.passwordError}>
                        <label>Repeat Password</label>
                        <Input value={this.state.repeatedPassword} type="password" onChange={(_e, { value }) => this.updateFields("repeatedPassword", value)} placeholder="Repeat password..." />
                    </Form.Field>
                    <Button type="button" content="Create" color="blue" onClick={() => this.submit()} loading={this.state.submitted} />
                </Form>
            </React.StrictMode>
        );
    }
}

export default UserCreation;