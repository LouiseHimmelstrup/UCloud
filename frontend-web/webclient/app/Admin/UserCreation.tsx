import * as React from "react";
import { Cloud } from "Authentication/SDUCloudObject";
import PromiseKeeper from "PromiseKeeper";
import { successNotification, defaultErrorHandler } from "UtilityFunctions";
import { UserCreationState, UserCreationField } from ".";
import { Input, Label, LoadingButton } from "ui-components";
import * as Heading from "ui-components/Heading";
import { connect } from "react-redux";
import { Dispatch } from "redux";
import { setActivePage } from "Navigation/Redux/StatusActions";
import { SidebarPages } from "ui-components/Sidebar";
import { MainContainer } from "MainContainer/MainContainer";

class UserCreation extends React.Component<UserCreationOperations, UserCreationState> {
    constructor(props: Readonly<UserCreationOperations>) {
        super(props);
        props.setActivePage();
        this.state = this.initialState;
    }

    private readonly initialState: UserCreationState = {
        promiseKeeper: new PromiseKeeper(),
        submitted: false,
        username: "",
        password: "",
        repeatedPassword: "",
        usernameError: false,
        passwordError: false
    };

    componentWillUnmount() {
        this.state.promiseKeeper.cancelPromises();
    }

    updateFields(field: UserCreationField, value: string) {
        const state = { ...this.state }
        state[field] = value;
        if (field === "username") state.usernameError = false;
        else if (field === "password" || field === "repeatedPassword") state.passwordError = false;
        this.setState(() => state);
    }

    submit(e: React.SyntheticEvent) {
        e.preventDefault();

        let usernameError = false;
        let passwordError = false;
        const { username, password, repeatedPassword } = this.state;
        if (!username) usernameError = true;
        if (!password || password !== repeatedPassword) passwordError = true;
        this.setState(() => ({ usernameError, passwordError }));
        if (!usernameError && !passwordError) {
            this.state.promiseKeeper.makeCancelable(
                Cloud.post("/auth/users/register", { username, password }, "")
            ).promise.then(f => {
                successNotification(`User '${username}' successfully created`);
                this.setState(() => this.initialState);
            }).catch(error => {
                const status = defaultErrorHandler(error);
                if (status == 400) {
                    this.setState(() => ({ usernameError: true }));
                }
            });
        }
    }

    render() {
        if (!Cloud.userIsAdmin) return null;

        const {
            usernameError,
            passwordError,
            username,
            password,
            repeatedPassword,
            submitted
        } = this.state;


        const header = (<><Heading.h1>User Creation</Heading.h1>
            <p>Admins can create new users on this page.</p></>)

        return (
            <MainContainer
                header={header}
                headerSize={120}
                main={<form onSubmit={e => this.submit(e)}>
                    <Label mb="1em">
                        Username
                        <Input
                            value={username}
                            color={usernameError ? "red" : "gray"}
                            onChange={({ target: { value } }) => this.updateFields("username", value)}
                            placeholder="Username..."
                        />
                    </Label>
                    <Label mb="1em">
                        Password
                        <Input
                            value={password}
                            type="password"
                            color={passwordError ? "red" : "gray"}
                            onChange={({ target: { value } }) => this.updateFields("password", value)}
                            placeholder="Password..."
                        />
                    </Label>
                    <Label mb="1em">
                        Repeat password
                        <Input
                            value={repeatedPassword}
                            type="password"
                            color={passwordError ? "red" : "gray"}
                            onChange={({ target: { value } }) => this.updateFields("repeatedPassword", value)}
                            placeholder="Repeat password..."
                        />
                    </Label>
                    <LoadingButton
                        type="submit"
                        content="Create user"
                        hovercolor="darkGreen"
                        color="green"
                        loading={submitted}
                    />
                </form>}
            />
        );
    }
}

interface UserCreationOperations {
    setActivePage: () => void
}

const mapDispatchToProps = (dispatch: Dispatch): UserCreationOperations => ({
    setActivePage: () => dispatch(setActivePage(SidebarPages.Admin))
});

export default connect(() => ({}), mapDispatchToProps)(UserCreation);