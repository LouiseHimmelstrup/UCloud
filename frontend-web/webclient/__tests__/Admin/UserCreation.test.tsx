import * as React from "react";
import UserCreation from "Admin/UserCreation";
import { create } from "react-test-renderer";
import { configure, mount } from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import { FormField, Form, Button } from "semantic-ui-react";
import PromiseKeeper from "PromiseKeeper";

configure({ adapter: new Adapter() });

describe("UserCreation", () => {
    test("Mount UserCreation component", () =>
        expect(create(<UserCreation />).toJSON()).toMatchSnapshot()
    );

    test("Update username field", () => {
        const userCreation = mount(<UserCreation />);
        userCreation.find(FormField).findWhere(it => it.props().label === "Username").find("input").simulate("change", { target: { value: "username" } });
        expect(userCreation.state()).toEqual({
            promiseKeeper: new PromiseKeeper(),
            submitted: false,
            username: "username",
            password: "",
            repeatedPassword: "",
            usernameError: false,
            passwordError: false
        });
    });

    test("Update password field", () => {
        const userCreation = mount(<UserCreation />);
        userCreation.find(FormField).findWhere(it => it.props().label === "Password").find("input").simulate("change", { target: { value: "password" } });
        expect(userCreation.state()).toEqual({
            promiseKeeper: new PromiseKeeper(),
            submitted: false,
            username: "",
            password: "password",
            repeatedPassword: "",
            usernameError: false,
            passwordError: false
        });
    });

    test("Update repeated password field", () => {
        const userCreation = mount(<UserCreation />);
        userCreation.find(FormField).findWhere(it => it.props().label === "Repeat password").find("input").simulate("change", { target: { value: "repeatWord" } });
        expect(userCreation.state()).toEqual({
            promiseKeeper: new PromiseKeeper(),
            submitted: false,
            username: "",
            password: "",
            repeatedPassword: "repeatWord",
            usernameError: false,
            passwordError: false
        });
    });

    // FIXME tests mounting and unmounting, but isn't this testing React as a framework?
    test("Mount and unmount component", () => {
        const userCreation = mount(<UserCreation />);
        userCreation.unmount();
        expect(userCreation.exists()).toBe(false);
    });

    test("Submit with missing username, causing errors to be rendered", () => {
        const userCreation = mount(<UserCreation />);
        userCreation.find(FormField).findWhere(it => it.props().label === "Password").find("input").simulate("change", { target: { value: "password" } });
        userCreation.find(FormField).findWhere(it => it.props().label === "Repeat password").find("input").simulate("change", { target: { value: "password" } });
        userCreation.find(Button).findWhere(it => it.props().content === "Create user").simulate("submit");
        expect(userCreation.state("usernameError")).toBe(true);
        expect(userCreation.state("passwordError")).toBe(false);
    });

    test("Submit with missing password fields, causing errors to be rendered", () => {
        const userCreation = mount(<UserCreation />);
        userCreation.find(FormField).findWhere(it => it.props().label === "Username").find("input").simulate("change", { target: { value: "username" } });
        userCreation.find(Button).findWhere(it => it.props().content === "Create user").simulate("submit");
        expect(userCreation.state("passwordError")).toBe(true);
        expect(userCreation.state("usernameError")).toBe(false);
    });

    test("Submit with non matching password fields, causing errors to be rendered", () => {
        const userCreation = mount(<UserCreation />);
        userCreation.find(FormField).findWhere(it => it.props().label === "Password").find("input").simulate("change", { target: { value: "passwordAlso" } });
        userCreation.find(FormField).findWhere(it => it.props().label === "Repeat password").find("input").simulate("change", { target: { value: "password" } });
        userCreation.find(Button).findWhere(it => it.props().content === "Create user").simulate("submit");
        expect(userCreation.state("usernameError")).toBe(true);
        expect(userCreation.state("passwordError")).toBe(true);
    });
});
