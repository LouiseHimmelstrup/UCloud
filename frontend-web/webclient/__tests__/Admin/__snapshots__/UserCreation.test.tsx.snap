// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserCreation Mount 1`] = `
<div
  className="ui container sixteen column grid"
>
  <div
    className="row"
  >
    <div
      className="sixteen wide column"
    >
      <div
        className="ui header"
      >
        <h1>
          User Creation
        </h1>
      </div>
      <p>
        Admins can create new users on this page.
      </p>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="sixteen wide column"
    >
      <form
        className="ui form"
        onSubmit={[Function]}
      >
        <div
          className="field"
        >
          <label>
            Username
          </label>
          <div
            className="ui input"
          >
            <input
              onChange={[Function]}
              placeholder="Username..."
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className="field"
        >
          <label>
            Password
          </label>
          <div
            className="ui input"
          >
            <input
              onChange={[Function]}
              placeholder="Password..."
              type="password"
              value=""
            />
          </div>
        </div>
        <div
          className="field"
        >
          <label>
            Repeat password
          </label>
          <div
            className="ui input"
          >
            <input
              onChange={[Function]}
              placeholder="Repeat password..."
              type="password"
              value=""
            />
          </div>
        </div>
        <button
          className="ui positive button"
          onClick={[Function]}
          role="button"
          type="submit"
        >
          <i
            aria-hidden="true"
            className="user plus icon"
          />
          Create user
        </button>
      </form>
    </div>
  </div>
</div>
`;
